### PR TITLE
Add a new attr indicates pub struct with private fields could be contructed externally

### LINKED ISSUE #126169: dead_code lint wrongly warns about public repr(C) structs with private fields but no constructors
### Code

```Rust
#[repr(C)] // <--
pub struct Foo {
    pub i: i16,
    align: i16
}
```

### Current output

```Shell
warning: field `align` is never read
 --> src/lib.rs:3:5
  |
1 | pub struct Foo {
  |            --- field in this struct
2 |     pub i: i16,
3 |     align: i16
  |     ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

### Desired output

```Shell
No warnings, or perhaps a different warning.
```

### Rationale and extra context

This warning is probably triggering because there's no constructor that allows a user to set the `align` field on struct `Foo`.  However, there is still a valid way for a user to construct it: `std::mem::zeroed()`.  The libc crate contains many structs that must be initialized that way.  libc could silence these warnings by `#[allow(dead_code)]`, but that's too broad a brush.  It would silence many much more serious warnings.  Could we please either revert to the previous behavior, or else create a new lint for this that can be separately silenced?

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.80.0-nightly (804421dff 2024-06-07)
binary: rustc
commit-hash: 804421dff5542c9c7da5c60257b5dbc849719505
commit-date: 2024-06-07
host: x86_64-unknown-freebsd
release: 1.80.0-nightly
LLVM version: 18.1.7
```

### Anything else?

This is a regression since `rustc 1.80.0-nightly (f67a1acc0 2024-06-01)`.  That version did not warn in such cases.

Downstream issue: https://github.com/rust-lang/libc/issues/3740

### PR DESCRIPTION
**Pull request by mu001999**
Fixes #126169

### CODE REVIEW

**Comment by compiler-errors** [comment]
https://github.com/rust-lang/rust/issues/126169#issuecomment-2166440393 notes #126289. Please add a test and/or support for that.

**Review by compiler-errors — COMMENTED** [review]
Also, this doesn't follow the usual process for adding new attributes to the compiler. 

This either needs to be FCP'd by T-lang as an insta-stable attribute, or it needs to be feature gated appropriately.

**Comment by compiler-errors** [comment]
@rustbot author

Please add feature gating. I also wonder if we should bikeshed this a bit first before adding it, like on i-rl-o.

File: compiler/rustc_passes/src/dead.rs [hunk 1]
```diff
@@ -59,21 +59,22 @@ fn struct_all_fields_are_public(tcx: TyCtxt<'_>, id: DefId) -> bool {
...
-        if is_positional
-            && tcx
-                .layout_of(tcx.param_env(field.did).and(field_type))
-                .map_or(true, |layout| layout.is_zst())
-        {
-            return true;
-        }
-        field.vis.is_public()
-    })
+    tcx.has_attr(id, sym::externally_constructed)
```
Comment (Soveu) [inline compiler/rustc_passes/src/dead.rs:62]: Is this a good place for this check? In function `check_item`, around line 838 you have a variable `may_construct_self` and in `visit_node` there is `unconditionally_treats_fields_as_live`

Comment (mu001999) [reply compiler/rustc_passes/src/dead.rs:62]: I think it's good, cause `struct_all_fields_are_public` is also used in `create_and_seed_worklist`.

Comment (mu001999) [reply compiler/rustc_passes/src/dead.rs]: Oh, seems checking it in `create_and_seed_worklist` is good enough. And I opened another PR which just checks the `repr(c)` or `repr(transparent)`, see #127104.

**Comment by compiler-errors** [comment]
I believe the other approach to this is #127104? If so, I'm gonna close this.

**Pull request closed without merging**